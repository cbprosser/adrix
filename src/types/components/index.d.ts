export * from './AdvancedSearchForm';
export * from './MTGStaticTypes';
export * from './MultipleCardDisplay';
export * from './SimpleCardSearchBar';
export * from './Symbol';
