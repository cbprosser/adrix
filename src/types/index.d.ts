export * from './axios';
export * from './components';
export * from './models';
export * from './redux';
export * from './sitemap';
