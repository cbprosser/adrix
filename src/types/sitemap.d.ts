import React from 'react';

export type SitePages = 'home' | 'results';
export type SiteURLS = '/' | '/search' | '/search/results';
export type SiteMapItem = {
  urls: SiteURLS[];
  Component: React.FC | (() => JSX.Element);
};

export type Sitemap = Record<SitePages, SiteMapItem>;
