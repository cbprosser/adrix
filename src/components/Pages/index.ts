export * from './AdvancedSearch';
export * from './Home';
export * from './Results';
