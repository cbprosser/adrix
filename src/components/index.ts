export * from './AppHeader';
export * from './CardDisplay';
export * from './Pages';
export * from './SimpleCardSearchBar';
