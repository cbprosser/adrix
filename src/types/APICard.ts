import { layouts, rarities, relatedComponents } from '../util';

/* eslint-disable camelcase */
export type APICard = APICore & APIGameplay & APIPrint;

export type APICore = {
  arena_id?: number;
  id: APIUUID;
  lang: string; // TODO: literalise
  mtgo_id?: number;
  mtgo_foil_id?: number;
  multiverse_ids?: number[];
  tcgplayer_id?: number;
  cardmarket_id?: number;
  object: 'card';
  oracle_id: APIUUID;
  prints_search_uri: APIURI;
  rulings_uri: APIURI;
  scryfall_uri: APIURI;
  uri: APIURI;
};

export type APIGameplay = {
  all_parts?: APIAllParts[];
  card_faces?: APICardFace[];
  cmc: number;
  color_identity: APIColors[];
  color_indicator?: APIColors[];
  colors?: APIColors[];
  edhrec_rank?: number;
  foil: boolean;
  hand_modifier?: string;
  keywords: string[];
  layout: APILayout;
  legalities: APILegalities;
  life_modifier?: string;
  loyalty?: string;
  mana_cost?: string;
  name: string;
  nonfoil: boolean;
  oracle_text?: string;
  oversized: boolean;
  power?: string;
  produced_mana?: APIColors[];
  reserved: boolean;
  toughness?: string;
  type_line: string;
};

export type APIPrint = {
  artist?: string;
  booster: boolean;
  border_color: string;
  card_back_id: APIUUID;
  collector_number: string;
  content_warning?: boolean;
  digital: boolean;
  flavor_name?: string;
  flavor_text?: string;
  frame_effects?: string[]; // TODO: literalise
  frame: string;
  full_art: boolean;
  games: string[]; // TODO: literalise
  highres_image: boolean;
  illustration_id?: APIUUID;
  image_status: string; // TODO: literalise
  image_uris?: APIImageURIS;
  prices: APIPrices;
  printed_name?: string;
  printed_text?: string;
  printed_type_line?: string;
  promo: boolean;
  promo_types?: string[];
  purchase_uris: APIPurchaseURIS;
  rarity: APIRarity;
  related_uris: APIRelatedURIS;
  released_at: APIDateString;
  reprint: boolean;
  scryfall_set_uri: APIURI;
  set_name: string;
  set_search_uri: APIURI;
  set_type: string;
  set_uri: APIURI;
  set: string;
  story_spotlight: boolean;
  textless: boolean;
  variation: boolean;
  variation_of?: APIUUID;
  watermark?: string;
  preview?: APIPreview;
};

export type APICardFace = {
  artist?: string;
  color_indicator?: string;
  colors?: APIColors[];
  flavor_text?: APIColors[];
  illustration_id?: APIUUID;
  image_uris?: APIImageURIS;
  loyalty?: string;
  mana_cost: string;
  name: string;
  object: string;
  oracle_text?: string;
  power?: string;
  printed_name?: string;
  printed_text?: string;
  printed_type_line?: string;
  toughness?: string;
  type_line: string;
  watermark?: string;
};

export type APIAllParts = {
  id: APIUUID;
  object: 'related_card';
  component: APIRelatedComponents;
  name: string;
  type_line: string;
  uri: APIURI;
};

export type APIRelatedURIS = {
  gatherer: APIURI;
  [key: string]: APIURI;
};

export type APIPurchaseURIS = {
  tcgplayer: APIURI;
  cardmarket: APIURI;
  cardhoarder: APIURI;
  [key: string]: APIURI;
};

export type APIPrices = {
  usd?: string;
  usd_foil?: string;
  eur?: string;
  eur_foil?: string;
  tix?: string;
};

export type APIImageURIS = {
  small?: APIURI;
  normal?: APIURI;
  large?: APIURI;
  png?: APIURI;
  art_crop?: APIURI;
  border_crop?: APIURI;
};

export type APIPreview = {
  previewed_at?: APIDateString;
  source_uri?: APIURI;
  source?: string;
};

export type APIColors = 'W' | 'U' | 'B' | 'R' | 'G';

export type APILegalities = {
  [key: string]: 'legal' | 'not legal' | 'restricted' | 'banned';
};

export type APIURI = string;

export type APIUUID = string;

export type APIDateString = string;

export type APILayout = typeof layouts[number];

export type APIRarity = typeof rarities[number];

export type APIRelatedComponents = typeof relatedComponents[number];
