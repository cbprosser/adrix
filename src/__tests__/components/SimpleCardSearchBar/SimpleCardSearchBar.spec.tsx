import { render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { SimpleCardSearchBar } from '../../../components';
import {
  initialCardsState,
  searchForCard,
  useAppSelector,
} from '../../../redux';
import { CardsState } from '../../../types';
import { MOCK, mockCard, mockInitialState } from '../../testUtils';

const mockDispatch = jest.fn();

jest.mock('../../../redux/thunks');

jest.mock('../../../redux/hooks', () => ({
  useAppSelector: jest.fn(),
  useAppDispatch: jest.fn(() => mockDispatch),
}));

jest.mock('axios');

const mockSelector = useAppSelector as jest.Mock;
const mockSearchForCard = searchForCard as unknown as jest.Mock;

describe('SimpleCardSearchBar Suite', () => {
  beforeEach(() => {
    jest.clearAllMocks();
    mockSelector.mockReturnValue(initialCardsState);
  });

  it('Should match the snapshot', () => {
    const view = render(<SimpleCardSearchBar />);

    expect(view.asFragment()).toMatchSnapshot();
  });

  it('Should set the proper state when a query is typed in', async () => {
    render(<SimpleCardSearchBar />);

    expect(screen.getByRole('textbox')).toBeDefined();

    await userEvent.type(screen.getByRole('textbox'), mockCard.name);

    expect(screen.getByRole<HTMLInputElement>('textbox').value).toEqual(
      mockCard.name,
    );
  });

  it('Should call for cards when the form is submitted', async () => {
    render(<SimpleCardSearchBar />);

    expect(screen.getByRole('textbox')).toBeDefined();

    expect(screen.getByRole<HTMLInputElement>('textbox').value).toEqual('');

    await userEvent.type(screen.getByRole('textbox'), mockCard.name);

    expect(screen.getByRole<HTMLInputElement>('textbox').value).toEqual(
      mockCard.name,
    );

    await userEvent.type(screen.getByRole('textbox'), '{enter}');

    expect(mockDispatch).toBeCalledWith(mockSearchForCard(mockCard.name));
  });

  it('Should be able to have its value controlled by props', () => {
    const view = render(<SimpleCardSearchBar value="" />);

    expect(screen.getByRole('textbox')).toBeDefined();

    expect(screen.getByRole<HTMLInputElement>('textbox').value).toEqual('');

    view.rerender(<SimpleCardSearchBar value={mockCard.name} />);

    expect(screen.getByRole<HTMLInputElement>('textbox').value).toEqual(
      mockCard.name,
    );
  });

  it('Should display a message when there is an error', () => {
    mockSelector.mockReturnValue({ message: MOCK.MESSAGE });
    render(<SimpleCardSearchBar />);

    expect(screen.getByText(MOCK.MESSAGE)).toBeDefined();
  });

  it('Should not display a message when there is an error and hideError is true', () => {
    mockSelector.mockReturnValue({
      ...initialCardsState,
      requests: { searchForCard: { message: MOCK.MESSAGE } },
    } as CardsState);
    render(<SimpleCardSearchBar hideError={true} />);

    expect(screen.queryByText(MOCK.MESSAGE)).toBeNull();
  });

  it('Should call the onChange callback when the value prop changes', async () => {
    const handleChange = jest.fn();
    render(<SimpleCardSearchBar onChange={handleChange} />);

    expect(screen.getByRole('textbox')).toBeDefined();

    await userEvent.type(screen.getByRole('textbox'), mockCard.name);

    expect(handleChange).toBeCalledWith(mockCard.name);
  });

  test('[COVERAGE] The inner function of useAppSelector', () => {
    render(<SimpleCardSearchBar />);

    expect(mockSelector.mock.calls[0][0](mockInitialState)).toEqual(
      initialCardsState.requests.searchForCard,
    );
  });
});
