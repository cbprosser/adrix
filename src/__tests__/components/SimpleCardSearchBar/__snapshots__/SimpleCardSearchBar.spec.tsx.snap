// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SimpleCardSearchBar Suite Should match the snapshot 1`] = `
<DocumentFragment>
  <form
    style="display: contents;"
  >
    <div
      class="MuiFormControl-root MuiTextField-root css-1u3bzj6-MuiFormControl-root-MuiTextField-root"
      style="width: 33.33333333333333vw; min-width: 300px;"
    >
      <label
        class="MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined MuiFormLabel-root MuiFormLabel-colorPrimary css-14s5rfu-MuiFormLabel-root-MuiInputLabel-root"
        data-shrink="false"
        for=":r0:"
        id=":r0:-label"
      >
        Enter a card name
      </label>
      <div
        class="MuiOutlinedInput-root MuiInputBase-root MuiInputBase-colorPrimary MuiInputBase-formControl css-9ddj71-MuiInputBase-root-MuiOutlinedInput-root"
      >
        <input
          aria-invalid="false"
          class="MuiOutlinedInput-input MuiInputBase-input css-1t8l2tu-MuiInputBase-input-MuiOutlinedInput-input"
          id=":r0:"
          type="text"
          value=""
        />
        <fieldset
          aria-hidden="true"
          class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
        >
          <legend
            class="css-1ftyaf0"
          >
            <span>
              Enter a card name
            </span>
          </legend>
        </fieldset>
      </div>
    </div>
  </form>
</DocumentFragment>
`;
